File: test/conformance/java/rows/P3.java
// test/conformance/java/rows/P3.java — the java/P3 cell of the hostile-input
// work set (docs/roadmap.sexp:3938), asserted for THIS leg at the tip.
//
// THE LAW, docs/FIXED-FORM-ALGORITHM.md:1686, §7 item 5:
//
//   "a byte-flip fuzz over the whole file, under a sanitizer, and it is not
//    optional — every offset is arithmetic over sizes a stranger wrote down,
//    so every byte, one bit at a time, is answered one of three ways and never
//    a fourth: a refusal by name, a `malformed` read, or a read that lands
//    values"
//
// THE VECTOR: the tree carries no fixed-form bytes in testdata, so the file is
// CONSTRUCTED here from the law — a lawful form-3 file of one GlobalSettings
// record, written by THIS build's own generated writer
// (build/tables-generated-java/examples/GlobalSettingsFixed.save), which is
// the same byte vector the reference's fuzz_case starts from
// (test/tables/fixedform_main.cpp:666-676). The reader under test is this
// build's IDENTITY plan over its own bytes, the JS leg's shape
// (test/js-tables/fixedform.mjs, hostileByteSweep).
//
// THE SANITIZER HALF is the language itself: an offset past the array this
// process owns throws IndexOutOfBoundsException, and an exception escaping a
// load that was asked a question IS the fourth answer — the same instrument
// the readers' oracle names (make/java.mk, tables-java-fuzz). Every byte of
// the file is flipped one bit at a time and the load must answer one of the
// three ways and never a fourth.
//
// Runs green with the generated code untouched and RED with one clamp removed
// from the generated reader (Step 3's control).
//
// RUN (from ./repo, the classpath make/java.mk gives the conformance driver):
//   java -cp build/conformance-java test/conformance/java/rows/P3.java

import java.nio.charset.StandardCharsets;
import tabledemo.GlobalSettingsFixed;
import tabledemo.TableFixed;

public final class P3 {
    private P3() {}

    private static int failures = 0;
    private static int assertions = 0;

    // One printed line per assertion.
    private static void check(boolean ok, String what) {
        assertions++;
        if (ok) {
            System.out.println("PASS: " + what);
        } else {
            System.out.println("FAILED: " + what);
            failures++;
        }
    }

    // A LANDED RECORD IS IN BOUND when every field the schema gives a bound to
    // sits inside it. GlobalSettings (tables/examples/Pack.schema) declares:
    //   tick_rate    uint32 = 60    min = 1, max = 240
    //   difficulty   Difficulty     (Easy/Normal/Hard, ordinals 0..3)
    //   build_note   string(48)
    //   spawn_delays [3]float32     (fixed three, no count on the wire)
    // The scatter's decode bounds are the pass the identity path owes (§7 item
    // 5); a hostile value that reaches the caller UNCLAMPED is the silently
    // wrong report this watches for.
    private static String issue(GlobalSettingsFixed.Value v) {
        if (v.tickRate < 1 || v.tickRate > 240) { return "tickRate " + v.tickRate; }
        if ((v.difficulty & 0xff) > 3) { return "difficulty " + (v.difficulty & 0xff); }
        if (v.buildNoteLength < 0 || v.buildNoteLength > 48) { return "buildNoteLength " + v.buildNoteLength; }
        return null;
    }

    public static void main(String[] args) {
        // THE LAWFUL VECTOR: one GlobalSettings record with every field set,
        // written by this build's own generated writer. The bytes are the
        // reader's own — form byte 3, the header, the layout, one record.
        final GlobalSettingsFixed.Value v = new GlobalSettingsFixed.Value();
        v.tickRate = 60;
        v.difficulty = 2;
        final byte[] note = "hostile".getBytes(StandardCharsets.UTF_8);
        System.arraycopy(note, 0, v.buildNote, 0, note.length);
        v.buildNoteLength = note.length;
        v.spawnDelays[0] = 1.5f;
        v.spawnDelays[1] = 2.5f;
        v.spawnDelays[2] = 3.5f;
        final byte[] file = new byte[GlobalSettingsFixed.measure(1)];
        check(GlobalSettingsFixed.save(new GlobalSettingsFixed.Value[] { v }, 1, file) == file.length,
                "P3 hostile bytes: the lawful record saves (" + file.length + " bytes)");

        // CONTROL 1's POSITIVE HALF, and it is the rule's own legitimate value:
        // the UNFLIPPED record must land, in bound, with no counter moved. A
        // sweep whose assertions fired on this would be testing something else.
        {
            final GlobalSettingsFixed.Value[] back = new GlobalSettingsFixed.Value[1];
            back[0] = new GlobalSettingsFixed.Value();
            final TableFixed.Report r = new TableFixed.Report();
            final int n = GlobalSettingsFixed.load(back, 1, file,
                    GlobalSettingsFixed.identityPlan(), new short[1024], GlobalSettingsFixed.image(), r);
            check(n == 1 && !r.malformed && !r.refused && r.clamped == 0,
                    "P3 hostile bytes: the lawful record reads clean (n=" + n
                            + ", refused=" + r.refused + ", malformed=" + r.malformed
                            + ", clamped=" + r.clamped + ")");
            check(issue(back[0]) == null,
                    "P3 hostile bytes: the lawful record lands every field in bound");
        }

        // THE SWEEP: every byte, one bit at a time. Each mutant is answered one
        // of three ways and never a fourth:
        //   refusal by name  n < 0 && refused && !malformed
        //   malformed        n < 0 && malformed && !refused
        //   lands values     n > 0, every landed field in bound, not refused
        // A throw escaping the load, a negative read that is neither refusal
        // nor malformed, or a landed field out of bound is the FOURTH.
        long refused = 0, malformed = 0, landed = 0, corrected = 0, divergences = 0;
        for (int at = 0; at < file.length; at++) {
            for (int bit = 0; bit < 8; bit++) {
                final byte[] hit = file.clone();
                hit[at] ^= (byte) (1 << bit);
                final GlobalSettingsFixed.Value[] back = new GlobalSettingsFixed.Value[1];
                back[0] = new GlobalSettingsFixed.Value();
                final TableFixed.Report r = new TableFixed.Report();
                int n;
                try {
                    n = GlobalSettingsFixed.load(back, 1, hit,
                            GlobalSettingsFixed.identityPlan(), new short[1024], GlobalSettingsFixed.image(), r);
                } catch (RuntimeException e) {
                    divergences++;
                    check(false, "P3 hostile bytes: byte " + at + " bit " + bit + " THREW " + e);
                    continue;
                }
                if (n < 0) {
                    if (r.refused && !r.malformed) {
                        refused++;
                    } else if (r.malformed && !r.refused) {
                        malformed++;
                    } else {
                        divergences++;
                        check(false, "P3 hostile bytes: byte " + at + " bit " + bit
                                + " answered a FOURTH way: n=" + n + ", refused=" + r.refused
                                + ", malformed=" + r.malformed);
                    }
                    continue;
                }
                landed++;
                if (r.clamped > 0) { corrected++; }
                if (r.refused) {
                    divergences++;
                    check(false, "P3 hostile bytes: byte " + at + " bit " + bit
                            + " returned records yet reported refused");
                }
                final String bad = issue(back[0]);
                if (bad != null) {
                    divergences++;
                    check(false, "P3 hostile bytes: byte " + at + " bit " + bit
                            + " landed OUT OF BOUND: " + bad);
                }
            }
        }

        // NOT VACUOUS: each of the three answers must actually have been SEEN,
        // and a correction must have moved a counter. A sweep that never
        // refused, never malformed, or never landed is a sweep over the wrong
        // file.
        check(refused > 0, "P3 hostile bytes: the sweep saw a named refusal (" + refused + ")");
        check(malformed > 0, "P3 hostile bytes: the sweep saw a malformed read (" + malformed + ")");
        check(landed > 0, "P3 hostile bytes: the sweep saw reads that landed values (" + landed + ")");
        check(corrected > 0, "P3 hostile bytes: the sweep exercised a correction that moved a counter (" + corrected + ")");
        check(divergences == 0, "P3 hostile bytes: " + divergences + " divergences, never a fourth");

        System.out.println("P3 hostile bytes: " + file.length + " bytes x 8 bits — " + refused
                + " refused by name, " + malformed + " malformed, " + landed
                + " records read (" + corrected + " with a correction), " + divergences + " divergences");
        if (failures == 0) {
            System.out.println("P3: " + assertions + " assertions, all green");
            System.exit(0);
        }
        System.out.println("P3: " + failures + " of " + assertions + " assertions FAILED");
        System.exit(1);
    }
}

File: test/conformance/js/rows/P3.mjs
// row js/P3 — "hostile bytes, sweep + sanitizer" (docs/roadmap.sexp js/P3;
// docs/FIXED-FORM-ALGORITHM.md §7 proof 5, §8 matrix row "byte-flip fuzz,
// sanitizer, three answers never a fourth").
//
// THE LAW, docs/FIXED-FORM-ALGORITHM.md:1686 (§7 proof 5):
//   "a byte-flip fuzz over the whole file, under a sanitizer, and it is not
//    optional — every offset is arithmetic over sizes a stranger wrote down,
//    so every byte, one bit at a time, is answered one of three ways and
//    never a fourth: a refusal by name, a `malformed` read, or a read that
//    lands values."
//
// The JS leg's fixed-form reader is the generated block reader
// (build/tables-generated-js/block/*Block.js, the same modules the
// conformance driver imports via SCHEMA_JS_GENERATED). Its answer set has no
// separate `malformed` verdict: bytes that are not this build's come back as
// a null handle — the refusal — and a valid image opens to a handle whose
// every row reads inside the bytes it was given — the read that lands values.
// The language's own sanitizer is the DataView's bounds check: a read past
// the view THROWS, so "no exception escaped a reader" IS "no read left the
// buffer". That escaping exception is the fourth answer this test says never
// happens — the reader that lets one out has stopped refusing and has not
// landed values, which is exactly the defect the leg's own fuzzer oracle
// calls "a read that escaped" (test/js-tables/main.mjs, fuzzOne).
//
// THE SWEEP. One bit at a time over every byte of a whole file: a valid
// block image, flipped once per mutant, opened at the file's own length with
// an aligned base (the caller's placement the driver also uses). Every mutant
// must be answered null, or a handle whose full row walk stays inside the
// extent. Any exception out of Open or out of the walk is a fourth answer and
// a failure.
//
// The image is built from the law's own datum — a block of THIS build, whose
// every offset is arithmetic over sizes the header declares — so the vector
// is the conformance corpus's own block_render.bin / block_padded.bin
// (testdata/wire/tables/), read through the SAME generated reader the driver
// drives on the block surfaces.

import { readFileSync } from "node:fs";
import { join, resolve } from "node:path";
import { pathToFileURL } from "node:url";

const generated = process.env.SCHEMA_JS_GENERATED ?? "build/tables-generated-js";
const load = (path) => import(pathToFileURL(resolve(generated, path)).href);

let failures = 0;
function assert(ok, what) {
  console.log((ok ? "OK   " : "FAIL ") + what);
  if (!ok) { failures++; }
}

// ---- the walk, over the descriptors and nothing else (the driver's dump
// and the leg's fuzzer both read exactly this way: every field of every row
// of every out-of-line array). A row read that would leave the bytes throws,
// which is the escaped read this test exists to say never happens.

function readField(view, at, f) {
  if (f.ElementRef !== null && f.ElemSize > 0) {
    for (const g of f.ElementRef().Fields) {
      const slots = g.IsArray ? g.ArrayBound : 1;
      for (let s = 0; s < slots; s++) { readField(view, at + g.Offset + s * g.ElemSize, g); }
    }
    return;
  }
  switch (f.ElemSize) {
    case 1: view.getUint8(at); return;
    case 2: view.getUint16(at, true); return;
    case 4: view.getUint32(at, true); return;
    case 8: view.getBigUint64(at, true); return;
    default: for (let i = 0; i < f.ElemSize; i++) { view.getUint8(at + i); }
  }
}

function walkBlock(block) {
  return (handle) => {
    const info = block.Type;
    for (const f of info.Fields) {
      if (!f.OutOfLine) { continue; }
      const offsetOf = Number(handle.View.getBigUint64(f.OffsetOfOffset, true));
      const count = handle.View.getUint32(f.CountOffset, true);
      const stride = handle.View.getUint32(f.StrideOffset, true);
      const rowInfo = f.ElementRef();
      for (let r = 0; r < count; r++) {
        const at = offsetOf + r * stride;
        for (const g of rowInfo.Fields) {
          const slots = g.IsArray ? g.ArrayBound : 1;
          for (let s = 0; s < slots; s++) { readField(handle.View, at + g.Offset + s * g.ElemSize, g); }
        }
      }
    }
  };
}

// ONE BYTE, ONE BIT: the sweep step. Every mutant is a single bit of a valid
// file reversed, opened at the file's own length on an aligned base. The three
// legal answers collapse to two verdicts in this reader's surface — null (the
// refusal) and a handle (a read that lands values) — and the sanitizer's own
// word for a read that left the buffer is the exception this returns as a
// fourth answer. Every mutant is counted, and the count is an assertion.
function sweep(open, walk, source) {
  let refused = 0;
  let opened = 0;
  let escaped = 0;
  let firstEscape = "";
  for (let i = 0; i < source.length; i++) {
    for (let b = 0; b < 8; b++) {
      const mutant = new Uint8Array(source.length);
      mutant.set(source);
      mutant[i] ^= 1 << b;
      let handle = null;
      try {
        handle = open(mutant);
      } catch (e) {
        escaped++;
        if (firstEscape === "") { firstEscape = "Open @byte " + i + " bit " + b + ": " + (e && e.message ? e.message : String(e)); }
        continue;
      }
      if (handle === null) { refused++; continue; }
      opened++;
      try {
        walk(handle);
      } catch (e) {
        escaped++;
        if (firstEscape === "") { firstEscape = "walk @byte " + i + " bit " + b + ": " + (e && e.message ? e.message : String(e)); }
      }
    }
  }
  return { refused, opened, escaped, firstEscape };
}

async function main() {
  const renderBlock = await load("block/RenderBlock.js");
  const paddedBlock = await load("block/PaddedBlock.js");

  const subjects = [
    {
      name: "block_render",
      file: "testdata/wire/tables/block_render.bin",
      open: (b) => renderBlock.RenderFrameBlock.Open(b),
      walk: walkBlock(renderBlock.RenderFrameBlock),
    },
    {
      name: "block_padded",
      file: "testdata/wire/tables/block_padded.bin",
      open: (b) => paddedBlock.PaddedFrameBlock.Open(b),
      walk: walkBlock(paddedBlock.PaddedFrameBlock),
    },
  ];

  for (const subject of subjects) {
    const source = new Uint8Array(readFileSync(subject.file));
    const clean = new Uint8Array(source.length);
    clean.set(source);
    assert(subject.open(clean) !== null, subject.name + ": the lawful image opens");
    const { refused, opened, escaped, firstEscape } = sweep(subject.open, subject.walk, source);
    assert(escaped === 0, subject.name + ": " + (source.length * 8) + " single-bit mutants (" + source.length +
      " bytes), " + refused + " refused, " + opened + " opened and walked whole inside the bytes they were given — " +
      (escaped === 0 ? "no exception escaped a reader" : firstEscape));
  }

  if (failures === 0) {
    console.log("row js/P3: every byte, one bit at a time, is answered by a refusal or a read that lands values — never a fourth (docs/FIXED-FORM-ALGORITHM.md:1686)");
  }
  process.exit(failures === 0 ? 0 : 1);
}

await main();